src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -856,11 +864,12 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher)
             changeWatcher.CacheExpires = DateTimeOffset.UtcNow.Add(cacheExpirationTime);
         }
 
-        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(IEnumerable<ConfigurationClient> clients, Func<ConfigurationClient, Task<T>> funcToExecute, CancellationToken cancellationToken = default)
+        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(IAsyncEnumerable<ConfigurationClient> clients, Func<ConfigurationClient, Task<T>> funcToExecute, CancellationToken cancellationToken = default)
         {
-            using IEnumerator<ConfigurationClient> clientEnumerator = clients.GetEnumerator();
+            IAsyncEnumerator<ConfigurationClient> clientEnumerator = clients.GetAsyncEnumerator(cancellationToken);
+            await using ConfiguredAsyncDisposable asyncDisposable = clientEnumerator.ConfigureAwait(false);
 
-            clientEnumerator.MoveNext();
+            await clientEnumerator.MoveNextAsync().ConfigureAwait(false);
```

Comment:
This can return false. Doesn't seem to be handled. If that does happen `clientEnumerator.Current` will be null.

Reply:
Umm.. it's been ensured that the `clients` won't be empty when this method is being called. But let's do the right thing here.

Reply:
Then it sounds like what should happen is that the caller passes an IAsyncEnumerator, rather than IAsyncEnumerable, and clientEnumerator.Current should be asserted to be non-null.

I saw the change you made and it's also problematic. ExecuteWithFailoverAsync should only return with success, or with an exception,